---
title: "Patel et al. 2024 revision, Nanopore analysis"
output:
  html_notebook: default
  pdf_document: default
editor_options: 
  chunk_output_type: inline
---

Read packages and methylation utilities:

```{r utils, message = F, warning = F}
library(tidyverse)
library(bsseq)
library(GenomicRanges)
library(khroma)
library(ggrepel)
library(parallel)

source("R/methylation_R_utils.R")
source("R/style.R")
```

Read the BSmoothed data in for GP5d WT and DNMTi+HDACi.

```{r import data, message = F, warning = F}
# gp5d_meth <- read.bismark("/data/kzkarttu/nanopore/GP5D/methylation/methylation_calls.GP5D.raw.cpg.mfreq.txt.gz")
# gp5d_inh_meth <- read.modkit("/data/kzkarttu/projects/Patel_2024_revision/data/Nanopore/GP5d_DNMTi_HDACi/methylation/GP5d_DNMTi_HDACi_bedMethyl_no_GCG.bed.gz")
# gp5d_meth_comb <- bsseq::combine(gp5d_meth, gp5d_inh_meth)

# gp5d_meth_comb_smooth <- BSmooth(gp5d_meth_comb, BPPARAM = p)
# saveRDS(gp5d_meth_comb_smooth, "data/bsmooth/GP5d_WT_inh_CpG_bsmoothed.rds")

gp5d_meth_comb_smooth <- readRDS("data/bsmooth/GP5d_WT_inh_CpG_bsmoothed.rds")
ltr12c_loci <- read_tsv("data/LTR12C/LTR12C_DNMT_HDAC_Up_bothstrand_GP5d.bed", col_names = c("seqnames", "start", "end", "name", "width", "strand"))

```

Tabulate methylation per each LTR12C locus in GP5d WT and GP5d DNMTi+HDACi.

```{r get_methylation}
ltr12c_meth_list <- getMeth(gp5d_meth_comb_smooth, regions = GRanges(ltr12c_loci), type = "smooth")

ltr12c_meth_all <- bind_rows(lapply(1:nrow(ltr12c_loci), function(i) {
  bind_cols(ltr12c_loci[i, ], as_tibble(ltr12c_meth_list[[i]]) %>%
    dplyr::rename(GP5d_WT = V1, GP5d_inh = V2))
}))

```

```{r plot_methylation_data}
plot_meth <- bind_rows(makeMethPlot(gp5d_meth_comb_smooth[, 1], GRanges(ltr12c_loci)),
  makeMethPlot(gp5d_meth_comb_smooth[, 2], GRanges(ltr12c_loci)),
  .id = "line"
) %>%
  mutate(
    line = if_else(line == 1, "GP5d WT", "GP5d DNMTi + HDACi"),
    dist = dist - median(dist)
  )
```

```{r agg_normalized_plot_data}
flank <- 1000

ltr12c_meth_loci <- bind_rows(mclapply(1:nrow(ltr12c_loci), mc.cores = 40, function(i) {
  coords <- bsseq::subsetByOverlaps(gp5d_meth_comb_smooth, GRanges(ltr12c_loci[i, ]) + flank)
  cov <- getCoverage(coords, type = "Cov", what = "perBase")
  m <- getCoverage(coords, type = "M", what = "perBase")

  bind_rows(
    ltr12c_loci[i, ] %>%
      bind_cols(., as_tibble(granges(coords)) %>% dplyr::select(-width, -strand, -end) %>% setNames(paste0("cpg_", names(.)))) %>%
      mutate(
        cov = cov[, 1],
        m = m[, 1]
      ),
    ltr12c_loci[i, ] %>%
      bind_cols(., as_tibble(granges(coords)) %>% dplyr::select(-width, -strand, -end) %>% setNames(paste0("cpg_", names(.)))) %>%
      mutate(
        cov = cov[, 2],
        m = m[, 2]
      ),
    .id = "cells"
  )
}))

ltr12c_meth_loci_all <- ltr12c_meth_loci %>%
  mutate(
    unmeth = cov - m,
    freq = m / (m + unmeth),
    flk = case_when(
      cpg_start < start ~ "flk_up",
      cpg_start > end ~ "flk_dn",
      cpg_start >= start & cpg_start <= end ~ "int"
    ),
    dist = case_when(
      flk == "int" ~ round((cpg_start - start) / (end - start), 3),
      flk == "flk_up" ~ round((cpg_start - start) / flank / 2, 3),
      flk == "flk_dn" ~ round(1 + (cpg_start - end) / flank / 2, 3),
    ),
  ) %>%
  na.omit() %>% 
  filter(cov >= 3)

ltr12c_meth_loci_list <- ltr12c_meth_loci_all %>%
  group_by(cells) %>%
  group_split(.keep = F)

ltr12c_meth_loci_ag <- lapply(1:2, function(i) {
  ltr12c_meth_loci_list[[i]] %>%
    group_by(dist) %>%
    dplyr::summarize(
      freq = mean(freq, na.rm = T),
    )
})

win <- .025

suppressWarnings(
ltr12c_meth_loci_ag_rollmeans <- lapply(1:2, function(i) {
  roll_range <- seq(
    from = min(ltr12c_meth_loci_ag[[i]]$dist) + win / 2,
    to = max(ltr12c_meth_loci_ag[[i]]$dist) - win / 2, by = .001
  )

  rollmeans <- numeric()

  for (center in roll_range) {
    dat_win <- ltr12c_meth_loci_ag[[i]][which(ltr12c_meth_loci_ag[[i]]$dist >= center - win &
      ltr12c_meth_loci_ag[[i]]$dist <= center + win), ]
    rollmean <- mean(dat_win$freq)
    rollmeans <- c(rollmeans, rollmean)
  }
  tibble(roll_range, rollmeans)
}))

```

Plot aggregated methylation difference in both GP5d cell lines, in a \pm 600bp window around the center of all the 500 LTR12C peaks.

```{r plot_agg, fig.width = 7.5, fig.height = 5}

bind_rows(ltr12c_meth_loci_ag_rollmeans, .id = "cells") %>% 
  mutate(cells = if_else(cells == 1, "GP5d WT", "GP5d DNMTi + HDACi")) %>% 
  ggplot(aes(x = roll_range, y = rollmeans, color = cells)) +
  geom_line(size = 1) +
  #geom_smooth(span = .01, method = "loess") +
  scale_color_manual(values = c("#ffba49", "#20a39e")) +
  guides(color = guide_legend(title = "Cells")) +
  scale_x_continuous(labels = c("-1kb", "Start", "End", "+1kb"), breaks = c(-0.4875, 0, 1, 1.4875)) +
  xlab("Position") +
  ylab("Methylation frequency") +
  geom_vline(xintercept = c(0, 1), color = "gray40", linetype = "dashed", alpha = .5) +
  theme(legend.position = c(.88, .91))

# ggsave("figures/240703_GP5d_wt_inh_agg_lineplot_length_normalized.pdf", width = 7.5, height = 5)
```

Also plotting data from FIMO of motifs at LTR12C, normalized by length, top 10 motifs by n of statistically significant hits.

```{r fimo_data, fig.width=6, fig.height=4}
dens_colors <- c("#001219", "#005F73", "#0A9396", "#94D2BD", "#E9D8A6", "#EE9B00", "#CA6702", "#BB3E03", "#AE2012", "#9B2226")

ltr12c_motifs <- read_tsv("/data/dpatel/GP5D/ChIP_seq/Seq_run_August_2022/SE_Analysis/Gain_Loss_of_KAP1_binding/Motif_analysis_upregulated_LTR12C/FIMO_GP5d_LTR12C/fimo.tsv") %>% 
  rename_with(~ tolower(gsub("-", "_", .x, fixed = TRUE))) # Renaming colums with dashed to underscores

ltr12c_motifs_filt <- ltr12c_motifs %>% 
  filter(q_value < 0.05) %>% 
  group_by(motif_alt_id) %>% 
  dplyr::summarize(n = n()) %>% 
  arrange(desc(n)) %>% 
  slice_head(n = 9) %>% 
  pull(motif_alt_id)
  
ltr12c_motifs %>% 
  filter(q_value < 0.05, motif_alt_id %in% ltr12c_motifs_filt) %>% 
  ggplot(aes(x = start, color = motif_alt_id, fill = motif_alt_id)) +
  geom_density(alpha = .25) +
  scale_color_muted() +
  scale_fill_muted()

ltr12c_motifs %>% 
  filter(q_value < 0.05, motif_alt_id %in% ltr12c_motifs_filt) %>% 
  mutate(ltr12c_start = as.numeric(str_remove(str_extract(sequence_name, ":\\d+"), ":")),
         ltr12c_end = as.numeric(str_remove(str_extract(sequence_name, "-\\d+$"), "-")),
         motif_start = ltr12c_start + start,
         norm_dist = (motif_start - ltr12c_start) / (ltr12c_end - ltr12c_start)) %>% 
  ggplot(aes(x = norm_dist, color = motif_alt_id, fill = motif_alt_id)) +
  geom_density(alpha = .3) +
  scale_color_muted() +
  scale_fill_muted()

```

Aggregated methylation line plot over all 500 LTR12C loci, not normalized for LTR12C length

```{r plot_methylation, fig.width = 7.5, fig.height = 5}
ggplot(plot_meth, aes(x = dist, y = freq, color = factor(line, ordered = T, levels = c("GP5d WT", "GP5d DNMTi + HDACi")))) +
  geom_line(size = 1) +
  xlab("Distance (bp)") +
  ylab("Methylation frequency") +
  scale_x_continuous(limits = c(-600, 600)) +
  scale_y_continuous(limits = c(0.3, 1)) +
  scale_color_manual(values = c("#ffba49", "#20a39e")) +
  guides(color = guide_legend(title = "Cells")) +
  theme(legend.position = c(.88, .91))

# ggsave("figures/240703_GP5d_wt_inh_agg_lineplot_non_normalized.pdf", width = 7.5, height = 5)
```

Comparing the extent of the difference in methylation between the two GP5d conditions, the table is sorted by the largest mean difference in methylation and the smallest adjusted p-value. "One" in this table means the GP5d WT and "Two" is GP5d DNMTi+HDACi.

```{r meth_difference_data}
gp5d_meth_comb_compare <- compareRegions(gp5d_meth_comb_smooth, 1, 2, GRanges(ltr12c_loci))

gp5d_meth_comb_compare %>%
  arrange(meandiff, adjusted.pval)
```

#### Volcano plot of mean differences in methylation

Plotting the difference in methylation between both conditions in a volcano plot. Top 6 largest mean differences are marked in the plot.

```{r meth_difference_data_plot, fig.width=6, fig.height=5}
volc_colors <- c("gray40", "#004488", "#BB5566")

gp5d_meth_comb_compare %>%
  mutate(
    lab = if_else(meandiff < -0.45, paste0(chr, ":", start, "-", end), NA),
    sig = case_when(
      adjusted.pval <= 0.05 & meandiff > 0 ~ "Significant (+)",
      adjusted.pval <= 0.05 & meandiff < 0 ~ "Significant (-)",
      .default = "Insignificant"
    )
  ) |> 
  select(meandiff, adjusted.pval, sig, lab) |> 
  mutate(neglog10_adjusted.pval = -log10(adjusted.pval)) |> 
  distinct() |> 
    write_tsv("data/source/Fig_4f_source_data.tsv")

gp5d_meth_comb_compare %>%
  mutate(
    lab = if_else(meandiff < -0.45, paste0(chr, ":", start, "-", end), NA),
    sig = case_when(
      adjusted.pval <= 0.05 & meandiff > 0 ~ "Significant (+)",
      adjusted.pval <= 0.05 & meandiff < 0 ~ "Significant (-)",
      .default = "Insignificant"
    )
  ) %>%
  ggplot(aes(x = meandiff, y = -log10(adjusted.pval), color = sig, label = lab)) +
  geom_point() +
  scale_colour_manual(values = volc_colors) +
  geom_hline(yintercept = -log10(0.05), color = "gray40", linetype = "dashed", alpha = .7) +
  geom_vline(xintercept = 0, color = "gray40", linetype = "dashed", alpha = .7) +
  xlab("Difference in mean methylation \nGP5d inhibited - GP5d WT") +
  ylab("-log10 (Adjusted p-value)") +
  geom_text_repel(direction = "x", color = "black") +
  guides(color = guide_legend(title = "Significance")) +
  theme(legend.position = c(.88, .88))

# ggsave("figures/240703_GP5d_wt_inh_methylation_volcano_plot.pdf", width = 6, height = 5)
```

#### Paired boxplot of the individual differences in methylation at upregulated GP5d DNMTi + HDACi LTR12C elements. 

```{r boxplot_LTR12C_boxplot, fig.width=4, fig.height=5}

gp5d_meth_comb_compare %>%
  pivot_longer(cols = c(average_one, average_two)) %>%
  mutate(
    name = factor(if_else(name == "average_one", "GP5d WT", "GP5d DNMTi + HDACi"), ordered = T, levels = c("GP5d WT", "GP5d DNMTi + HDACi")),
    id = rep(1:500, each = 2, length.out = 1000)
  ) %>% # Making an ID column for paired boxplot
  ggplot(aes(x = name, y = value, fill = name)) +
  geom_boxplot(aes(group = name), outlier.shape = NA, alpha = .75) +
  # geom_jitter(width = 0.3, size = .2, seed = 1234) +
  geom_point(size = .2, alpha = .25, position = position_jitter(seed = 1234, width = .38)) +
  geom_line(aes(group = factor(id)), size = .05, alpha = .05, position = position_jitter(seed = 1234, width = .38)) +
  scale_fill_manual(values = c("#ffba49", "#20a39e")) +
  theme(axis.title.x = element_blank(),
        legend.position = "none") +
  ylab("Methylation") 

# ggsave("figures/240703_GP5d_wt_inh_LTR12C_methylation_boxplot.pdf", width = 4, height = 5)
```

#### Same plot as above but for all downregulated TE subfamilies in GP5d DNMTi + HDACi.

```{r plot_boxplot_all_down, fig.width=4, fig.height=5}
GP5d_inh_all_down <- read_tsv("data/TEs/GP5d_DNMTi_HDACi_Downregulated.bed", col_names = c("seqnames", "start", "end", "name", "family", "class"))

gp5d_meth_comb_compare_all_down <- compareRegions(gp5d_meth_comb_smooth, 1, 2, GRanges(GP5d_inh_all_down))

gp5d_meth_comb_compare_all_down %>%
  pivot_longer(cols = c(average_one, average_two)) %>%
  mutate(
    name = factor(if_else(name == "average_one", "GP5d WT", "GP5d DNMTi + HDACi"), ordered = T, levels = c("GP5d WT", "GP5d DNMTi + HDACi")),
    id = rep(1:(1570 / 2), each = 2, length.out = 1570)
  ) %>% # Making an ID column for paired boxplot
  ggplot(aes(x = name, y = value, fill = name)) +
  geom_boxplot(aes(group = name), outlier.shape = NA, alpha = .75) +
  # geom_jitter(width = 0.3, size = .2, seed = 1234) +
  geom_point(size = .2, alpha = .25, position = position_jitter(seed = 1234, width = .38)) +
  geom_line(aes(group = factor(id)), size = .05, alpha = .05, position = position_jitter(seed = 1234, width = .38)) +
  scale_fill_manual(values = c("#ffba49", "#20a39e")) +
  theme(axis.title.x = element_blank(),
        legend.position = "none") +
  ylab("Methylation") 

# ggsave("figures/240703_GP5d_wt_inh_all_down_methylation_boxplot.pdf", width = 4, height = 5)
```

#### Same plot as above but for all UPregulated TE subfamilies in GP5d DNMTi + HDACi.

```{r plot_boxplot_all_up, fig.width=4, fig.height=5}
gp5d_inh_all_up <- read_tsv("data/TEs/GP5d_DNMTi_HDACi_Upregulated.bed", col_names = c("seqnames", "start", "end", "name", "family", "class"))
gp5d_meth_comb_compare_all_up <- compareRegions(gp5d_meth_comb_smooth, 1, 2, GRanges(gp5d_inh_all_up))

gp5d_meth_comb_compare_all_up %>%
  pivot_longer(cols = c(average_one, average_two)) %>%
  mutate(
    name = factor(if_else(name == "average_one", "GP5d WT", "GP5d DNMTi + HDACi"), ordered = T, levels = c("GP5d WT", "GP5d DNMTi + HDACi")),
    id = rep(1:(nrow(.) / 2), each = 2, length.out = nrow(.))
  ) %>% # Making an ID column for paired boxplot
  ggplot(aes(x = name, y = value, fill = name)) +
  geom_boxplot(aes(group = name), outlier.shape = NA, alpha = .75) +
  # geom_jitter(width = 0.3, size = .2, seed = 1234) +
  geom_point(size = .2, alpha = .25, position = position_jitter(seed = 1234, width = .38)) +
  geom_line(aes(group = factor(id)), size = .05, alpha = .05, position = position_jitter(seed = 1234, width = .38)) +
  scale_fill_manual(values = c("#ffba49", "#20a39e")) +
  theme(axis.title.x = element_blank(),
        legend.position = "none") +
  ylab("Methylation") 

# ggsave("figures/240703_GP5d_wt_inh_all_up_methylation_boxplot.pdf", width = 4, height = 5)
```

#### Plotting the differences by TE class in an XY plot.

```{r plot_xy_by_TE_class, fig.width=6, fig.height=4.5}
xy_colors <- c("#003049", "#D62828", "#F77F00", "#FCBF49", "gray40")

gp5d_meth_comb_compare_all_up  %>% 
  mutate(id = paste0(chr, ":", start, "-", end)) %>% 
  left_join(., GP5d_inh_all_up %>% mutate(id = paste0(seqnames, ":", start, "-", end)) %>% dplyr::select(-seqnames, -start, -end), by = "id") %>% 
  mutate(clr = if_else(adjusted.pval <= 0.05, class, "Insignificant"),
         clr = factor(clr, levels = c("LINE", "SINE", "LTR", "DNA", "Insignificant")),
         sz = factor(if_else(clr != "Insignificant", 1, 2))) |> 
  select(average_one, average_two, clr, class, sz) |> 
  write_tsv("data/source/Supp_fig_8a_source_data.tsv")

gp5d_meth_comb_compare_all_up  %>% 
  mutate(id = paste0(chr, ":", start, "-", end)) %>% 
  left_join(., GP5d_inh_all_up %>% mutate(id = paste0(seqnames, ":", start, "-", end)) %>% dplyr::select(-seqnames, -start, -end), by = "id") %>% 
  mutate(clr = if_else(adjusted.pval <= 0.05, class, "Insignificant"),
         clr = factor(clr, levels = c("LINE", "SINE", "LTR", "DNA", "Insignificant")),
         sz = factor(if_else(clr != "Insignificant", 1, 2))) %>% 
  ggplot(aes(x = average_one, y = average_two, color = clr, fill = clr)) +
  facet_wrap(~class, nrow = 2) + 
  geom_point(aes(size = sz, alpha = sz), shape = 21) +
  geom_abline(color = "gray40", alpha = .5, linetype = "dashed") +
  scale_color_manual(values = xy_colors) +
  scale_fill_manual(values = xy_colors) +
  scale_size_manual(values = c(1, .3)) +
  scale_alpha_manual(values = c(.75, .2)) +
  xlab("GP5d WT\nmethylation frequency") +
  ylab("GP5d DNMTi + HDACi\nmethylation frequency") +
  theme(legend.position = "none") +
  scale_x_continuous(breaks = c(0, 0.25, 0.5, 0.75, 1), limits = c(0, 1), labels = c("0", "0.25", "0.5", "0.75", "1")) +
  scale_y_continuous(breaks = c(0, 0.25, 0.5, 0.75, 1), limits = c(0, 1), labels = c("0", "0.25", "0.5", "0.75", "1"))

# ggsave("figures/240703_GP5d_wt_inh_all_up_methylation_xyplot_by_class.pdf", width = 6, height = 4.5)
```
